fix(core): do not invoke jasmine done callback multiple times with `waitForAsync`

devversion opened the pull request:
See individual commits. This evolved from the simple `karma-jasmine` update (you would think "easy" 😛)

The `zone.js` fix is necessary for us to update `karma-jasmine` where multiple `done` invocations result in runtime errors.

angular-robot committed:
build: update dependency karma-jasmine to v5

| datasource | package       | from  | to    |
| ---------- | ------------- | ----- | ----- |
| npm        | karma-jasmine | 4.0.2 | 5.1.0 |

devversion committed:
fix(core): do not invoke jasmine `done` callback multiple times with `waitForAsync`

Currently tests written using `waitForAsync` would be prone to Jasmine
warnings or errors (depending on the version) for tests incorrectly
invoking asynchronous jasmine `done` callbacks multiple times.

This can happen because the async test zone logic schedules the
`done` callback to be called using `setTimeout`, but this could
be invoked multiple times, causing multiple `done` invocations to
be scheduled. Most of the issues have been resolved with #45025,
but it does not solve the case of multiple tasks finished and callbacks
being scheduled.

Technically, the current logic is built in way that _should_ result in
`_finishCallbackIfDone` and eventually the `done` callback to be invoked
at maximium once. This is unfortunately not the case in some rather
advanced/unexpected scenarios (like our AngularJS upgrade tests) where
the scenario is the following (and microtasks from before the actual
`waitForAsync` spec are still completing -- which is valid):

```
1. A test `beforeEach` schedules a microtask in the ProxyZone.
2. An actual empty `it` spec executes in the AsyncTestZone` (using e.g. `waitForAsync`).
3. The `onInvoke` invokes `_finishCallbackIfDone` because the spec runs synchronously.
4. We wait the scheduled timeout (see below) to account for unhandled promises.
5. The microtask from (1) finishes and `onHasTask` is invoked.

--> We register a second `_finishCallbackIfDone` even though we have scheduled a timeout.
--> we execute the `done` callback twice because the async zone spec state is "stable"
```

JoostK commented:
This probably fixes #45476.

devversion committed:
test: update tests to account for `karma-jasmine` v5.0.0

Karma jasmine updated the `jasmine-core` dependency. Jasmine is now more
strict when:

* The done callback is invoked multiple times
* The done callback is used, while a promise is also returned
* The done callback is treated as error when e.g. a number is returned
  as first argument. This was the case with `requestAnimationFrame`.

jessicajaniuk commented on packages/zone.js/test/browser/browser.spec.ts:
```diff
@@ -3398,145 +3398,140 @@ describe('Zone', function() {
-              'MSPointerOver': 'pointerover',
-              'MSPointerUp': 'pointerup'
-            };
+  if (getIEVersion() === 11) {
```
Can we not eliminate these tests entirely since we no longer support Internet Explorer?

devversion replied:
yeah, I was thinking the same, but:

* This would be better in a separate PR/ is out of scope here
* It's unclear (at least to me, reading the docs) if zonejs has its own support guarantees)

jessicajaniuk replied:
We should maybe add an issue for this work, then. Otherwise, it'll get lost / forgotten.

devversion replied:
Agreed, created: https://github.com/angular/angular/issues/46773

devversion committed:
test: fix `fake-async-test` spec to properly test prefixed animation frame functions

In the fake async test for `zone.js` there is some logic to test various
prefixes of the request animation frame functions. This logic does not
have any effect currently, except for generating the same test at
maximum three times.

This commit fixes the test to actually test what it meant to do
initially.

josephperrott reviewed (approved):
LGTM

jessicajaniuk commented on packages/zone.js/test/browser/requestAnimationFrame.spec.ts:
```diff
@@ -6,53 +6,52 @@
+        afterEach(() => {
+          (<any>jasmine).DEFAULT_TIMEOUT_INTERVAL = originalTimeout;
+        });
+        const rAF = (window as any)[fnName];
```
Can we make this a clearer, non-abbreviated name?

jessicajaniuk reviewed (approved):
LGTM

JiaLiPassion commented:
I just removed the `action: merge` label, since I still need a little time to review, I will update the result asap.

JiaLiPassion commented:
@devversion , thank you, LGTM, I will continue to research the `separate ProxyZone` solutions.

devversion commented:
For additional context: This PR caught one edge case (as documented in the async test zone implementation), caused by the fact that the Proxy zone is shared between all Jasmine specs (and before-each etc.). This is the behavior for quite some time, and where existing tests & users could rely on.

There is the idea of having its own proxy zone for every jasmine spec, so that they would not leak through in between (and therefore would not affect the test async zone spec). This might break existing tests and modifies some of the semantics in ZoneJS/Jasmine/Angular (tests potentially shouldn't rely on this -- this is something too figure out). @JiaLiPassion is exploring this in a separate PR.

This PR fixes the Jasmine issue without intending to change the semantics, so we will go with that as a simple fix.

devversion commented:
Okay, so I have passing TGP again: tap/OCL:460707153:BASE:460901104:1657798948408:ca978ff4

devversion committed:
feat(zone.js): include zone name when sync-test zone reports tasks

The sync-test zone is used in e.g. `describe` to raise an error when
there is asynchronous code scheduled in describe blocks. This commit
includes the zone name in such thrown errors to allow for us to include
the jasmine describe name in the error. This will be wired up in the
jasmine zonejs patches separately.

devversion committed:
feat(zone.js): include jasmine `describe` block name when raising unexpected task error

As mentioned in the previous commit that ensured that the Zone name is
included in errors raised by the `SyncTestZoneSpec`, we can now include
the Jasmine describe block descriptions in such errors.

Errors can often happen when users accidentally try to set up Angular
tests without an `it` block. Resulting in errors where it's not clear
at all which describe block (of potentially a large repository) is
involved:

```
 An error was thrown in afterAll
  error properties: Object({ originalStack: 'Error: Cannot call XX from within a sync test.
      at new ZoneAwareError (packages/zone.js/test/browser_test_rollup.umd.js:98:37)
      at e.onScheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:158:196)
      at e.scheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:7529)
      at t.scheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:3539)
      at t.scheduleMicroTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:3791)
      at r.execute (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:166:4372)
      at queueRunnerFa ...
      at <Jasmine>
```

We now include the describe block description in the error, so that it
is easier to figure out the location of the culprit code.

devversion committed:
fixup! feat(zone.js): include jasmine `describe` block name when raising unexpected task error

ds

devversion committed:
test: update tests to not run jasmine `done` function in sync-test zone from `describe` block

There are some ZoneJS tests that fork the zone from the `describe` block
for testing the zone patching. This does cause the Jasmine `done`
function later in `it` specs to be invoked in the sync-test zone from
the original `describe` block. The `done` implementation now has
changed with the Karma Jasmine update and breaks because it now causes
tasks to be scheduled.

It is conceptually incorrect/invalid to take the describe sync zone and
run test logic with that sync zone.

```
 An error was thrown in afterAll
  error properties: Object({ originalStack: 'Error: Cannot call jasmine.execute().forceTask from within a sync test (syncTestZone for jasmine.describe#FileReader).
      at new ZoneAwareError (packages/zone.js/test/browser_test_rollup.umd.js:98:37)
      at e.onScheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:158:196)
      at e.scheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:7529)
      at t.scheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:3539)
      at t.scheduleMicroTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:3791)
      at r.execute (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:166:4312)
      at queueRunnerFa ...
      at <Jasmine>
```

devversion commented:
[Another passing TGP.](https://fusion2.corp.google.com/presubmit/tap/460707153/OCL:460707153:BASE:461181249:1657888374216:d0e0bc15/targets)
